File: data/src/main/java/org/cryptomator/data/cloud/pcloud/PCloudNodeFactory.java
package org.cryptomator.data.cloud.pcloud;

import com.pcloud.sdk.RemoteEntry;
import com.pcloud.sdk.RemoteFile;
import com.pcloud.sdk.RemoteFolder;

import org.cryptomator.util.Optional;

class PCloudNodeFactory {

	public static PCloudFile file(PCloudFolder parent, RemoteFile file) {
		return new PCloudFile(parent, file.name(), getNodePath(parent, file.name()), Optional.ofNullable(file.size()), Optional.ofNullable(file.lastModified()));
	}

	public static PCloudFile file(PCloudFolder parent, String name, Optional<Long> size) {
		return new PCloudFile(parent, name, getNodePath(parent, name), size, Optional.empty());
	}

	public static PCloudFile file(PCloudFolder parent, String name, Optional<Long> size, String path) {
		return new PCloudFile(parent, name, path, size, Optional.empty());
	}

	public static PCloudFolder folder(PCloudFolder parent, RemoteFolder folder) {
		return new PCloudFolder(parent, folder.name(), getNodePath(parent, folder.name()));
	}

	public static PCloudFolder folder(PCloudFolder parent, String name) {
		return new PCloudFolder(parent, name, getNodePath(parent, name));
	}

	public static PCloudFolder folder(PCloudFolder parent, String name, String path) {
		return new PCloudFolder(parent, name, path);
	}

	public static String getNodePath(PCloudFolder parent, String name) {
		return parent.getPath() + "/" + name;
	}

	public static PCloudNode from(PCloudFolder parent, RemoteEntry remoteEntry) {
		if (remoteEntry instanceof RemoteFile) {
			return file(parent, remoteEntry.asFile());
		} else {
			return folder(parent, remoteEntry.asFolder());
		}
	}

}

File: data/src/main/java/org/cryptomator/data/cloud/pcloud/PCloudNodeFactory.kt
package org.cryptomator.data.cloud.pcloud

import com.pcloud.sdk.RemoteEntry
import com.pcloud.sdk.RemoteFile
import com.pcloud.sdk.RemoteFolder

internal object PCloudNodeFactory {

	fun file(parent: PCloudFolder, file: RemoteFile): PCloudFile {
		return PCloudFile(parent, file.name(), getNodePath(parent, file.name()), file.size(), file.lastModified())
	}

	fun file(parent: PCloudFolder, name: String, size: Long?): PCloudFile {
		return PCloudFile(parent, name, getNodePath(parent, name), size, null)
	}

	@JvmStatic
	fun file(parent: PCloudFolder, name: String, size: Long?, path: String): PCloudFile {
		return PCloudFile(parent, name, path, size, null)
	}

	fun folder(parent: PCloudFolder, folder: RemoteFolder): PCloudFolder {
		return PCloudFolder(parent, folder.name(), getNodePath(parent, folder.name()))
	}

	fun folder(parent: PCloudFolder, name: String): PCloudFolder {
		return PCloudFolder(parent, name, getNodePath(parent, name))
	}

	@JvmStatic
	fun folder(parent: PCloudFolder?, name: String, path: String): PCloudFolder {
		return PCloudFolder(parent, name, path)
	}

	fun getNodePath(parent: PCloudFolder, name: String): String {
		return parent.path + "/" + name
	}

	@JvmStatic
	fun from(parent: PCloudFolder, remoteEntry: RemoteEntry): PCloudNode {
		return if (remoteEntry is RemoteFile) {
			file(parent, remoteEntry.asFile())
		} else {
			folder(parent, remoteEntry.asFolder())
		}
	}
}
